tiledb/api/src/filter.rs

```diff
@@ -54,26 +274,139 @@ impl<'ctx> Filter<'ctx> {
         Filter { context, raw }
     }
 
-    fn create(
+    pub fn create(
         context: &'ctx Context,
-        filter_type: ffi::FilterType,
+        filter_data: FilterData,
     ) -> TileDBResult<Self> {
+        let c_context = context.capi();
         let mut c_filter: *mut ffi::tiledb_filter_t = out_ptr!();
-        let ftype = filter_type as u32;
+        let ftype = filter_data.capi_enum() as u32;
         let res = unsafe {
-            ffi::tiledb_filter_alloc(context.capi(), ftype, &mut c_filter)
+            ffi::tiledb_filter_alloc(c_context, ftype, &mut c_filter)
         };
-        if res == ffi::TILEDB_OK {
-            Ok(Filter {
-                context,
-                raw: RawFilter::Owned(c_filter),
-            })
-        } else {
-            Err(context.expect_last_error())
+        if res != ffi::TILEDB_OK {
+            return Err(context.expect_last_error());
         }
+
+        let raw = RawFilter::Owned(c_filter);
+
+        match filter_data {
+            FilterData::None => (),
+            FilterData::BitShuffle { .. } => (),
+            FilterData::ByteShuffle { .. } => (),
+            FilterData::BitWidthReduction { max_window } => {
+                if let Some(max_window) = max_window {
+                    let c_size = max_window as std::ffi::c_uint;
+                    Self::set_option(
+                        context,
+                        *raw,
+                        ffi::FilterOption::BIT_WIDTH_MAX_WINDOW,
+                        c_size,
+                    )?;
+                }
+            }
+            FilterData::Checksum(ChecksumType::Md5) => (),
+            FilterData::Checksum(ChecksumType::Sha256) => (),
+            FilterData::Compression(CompressionData {
+                level,
+                reinterpret_datatype,
+                ..
+            }) => {
+                if let Some(level) = level {
+                    let c_level = level as std::ffi::c_int;
+                    Self::set_option(
+                        context,
+                        *raw,
+                        ffi::FilterOption::COMPRESSION_LEVEL,
+                        c_level,
+                    )?;
+                }
+                if let Some(reinterpret_datatype) = reinterpret_datatype {
+                    let c_datatype =
+                        reinterpret_datatype.capi_enum() as std::ffi::c_uchar;
```

Comment:
is `std::ffi::c_uchar` correct here? I would have expected `u32` or `std::ffi::c_uint`.

Reply:
Surprisingly yes - it expects just one byte inside of libtiledb

Reply:
Well that's awkward. In a note to future selves, I tracked this down to the fact that most of our uses of Datatype refelect the public C API `typedef enum { ... } tiledb_datatype_t` definition which is apparently backed by a `u32` according to bindgen. However, the `Filter::set_option` pipes the `void*` all the way to `CompressionFilter::set_option_impl` which is casting it to a `tiledb::sm::Datatype` which is `enum class Datatype : uint8_t { ... }` definition which means it really does want a single byte ini this case.